pkg/cloudprovider/provider/vsphere/helper.go

```diff
@@ -218,13 +211,80 @@ func createClonedVM(ctx context.Context, vmName string, config *Config, session
 	return virtualMachine, nil
 }
 
-func uploadAndAttachISO(ctx context.Context, session *Session, vmRef *object.VirtualMachine, localIsoFilePath, datastoreName string) error {
-	datastore, err := session.Finder.Datastore(ctx, datastoreName)
-	if err != nil {
-		return fmt.Errorf("failed to get datastore: %v", err)
+func resolveDatastoreRef(ctx context.Context, config *Config, session *Session, vm *object.VirtualMachine, folder *object.Folder, cloneSpec *types.VirtualMachineCloneSpec) (*types.ManagedObjectReference, error) {
+	// Based on https://github.com/vmware/govmomi/blob/v0.22.1/govc/vm/clone.go#L358
+	if config.DatastoreCluster != "" && config.Datastore == "" {
+		klog.Infof("Choosing initial datastore placement for vm %s from datastore cluster %s", vm.Name(), config.DatastoreCluster)
+		storagePod, err := session.Finder.DatastoreCluster(ctx, config.DatastoreCluster)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get datastore cluster: %v", err)
+		}
+
+		// Build pod selection spec from config spec
+		podSelectionSpec := types.StorageDrsPodSelectionSpec{
+			StoragePod: types.NewReference(storagePod.Reference()),
+		}
+
+		// Get the virtual machine reference
+		vmref := vm.Reference()
+
+		// TODO(irozzo) moveAllDiskBackingsAndConsolidate does not seem to work with RecommendDatastore,
+		// try to better understand the reason and the implications.
+		// https://code.vmware.com/docs/4206/vsphere-web-services-api-reference/doc/vim.vm.RelocateSpec.DiskMoveOptions.html
+		cloneSpec.Location.DiskMoveType = string(types.VirtualMachineRelocateDiskMoveOptionsMoveAllDiskBackingsAndDisallowSharing)
+		// Build the placement spec
+		storagePlacementSpec := types.StoragePlacementSpec{
+			Folder:           types.NewReference(folder.Reference()),
+			Vm:               &vmref,
+			CloneName:        vm.Name(),
+			CloneSpec:        cloneSpec,
+			PodSelectionSpec: podSelectionSpec,
+			Type:             string(types.StoragePlacementSpecPlacementTypeClone),
+		}
+
+		// Get the storage placement result
+		storageResourceManager := object.NewStorageResourceManager(session.Client.Client)
+		result, err := storageResourceManager.RecommendDatastores(ctx, storagePlacementSpec)
+		if err != nil {
+			return nil, fmt.Errorf("error occurred while getting storage placement recommendation: %v", err)
+		}
+
+		// Get the recommendations
+		recommendations := result.Recommendations
+		if len(recommendations) == 0 {
```

Comment:
If there was no recommended datastore in this cluster, can't we just pick a random one and report a warning error, instead of completely aborting the task?

Reply:
I think it's better to fail in such circumstances, because it's an unexpected condition. Note that this code is inspired by the[ vmware command line tool](https://github.com/vmware/govmomi/blob/e179aa206e05606aeb5b1fde9c79a6613a0e048d/govc/vm/clone.go#L385) , as I'm not a vSphere expert this was my reference.

Reply:
Please check with the customer or with kamil about this!

Reply:
It seems that failing to get a recommendation is an error condition and vsphere. Let's treat it as such unless we find a compelling reason to ignore the failure.